pytorch/mel2samp_onehot.py

```diff
@@ -0,0 +1,136 @@
+#  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+#  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+#  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# 
+# *****************************************************************************
+"""
+Generating pairs of mel-spectrograms and original audio
+"""
+import argparse
+import json
+import os
+import random
+import torch
+import torch.utils.data
+import sys
+
+import utils
+
+# We're using the audio processing from TacoTron2 to make sure it matches
+sys.path.insert(0, 'tacotron2')
+from tacotron2.layers import TacotronSTFT
+
+class Mel2SampOnehot(torch.utils.data.Dataset):
+    """
+    This is the main class that calculates the spectrogram and returns the
+    spectrogram, audio pair.
+    """
+    def __init__(self, training_files, segment_length, mu_quantization,
+                 filter_length, hop_length, win_length, sampling_rate):
+        audio_files = utils.files_to_list(training_files)
+        self.audio_files = audio_files
+        random.seed(1234)
+        random.shuffle(self.audio_files)
+        
+        self.stft = TacotronSTFT(filter_length=filter_length,
+                                    hop_length=hop_length,
+                                    win_length=win_length,
+                                    sampling_rate=sampling_rate)
+        
+        self.segment_length = segment_length
+        self.mu_quantization = mu_quantization
+        self.sampling_rate = sampling_rate
+
+    def get_mel(self, audio):
+        audio_norm = audio / utils.MAX_WAV_VALUE
+        audio_norm = audio_norm.unsqueeze(0)
+        audio_norm = torch.autograd.Variable(audio_norm, requires_grad=False)
+        melspec = self.stft.mel_spectrogram(audio_norm)
+        melspec = torch.squeeze(melspec, 0)
+        return melspec
+    
+    def __getitem__(self, index):
+        # Read audio
+        filename = self.audio_files[index]
+        audio, sampling_rate = utils.load_wav_to_torch(filename)
+        if sampling_rate != self.sampling_rate:
+            raise ValueError("{} {} SR doesn't match target {} SR".format(
+                sampling_rate, self.sampling_rate))
+        
+            # Take segment
+        if audio.size(0) >= self.segment_length:
+            max_audio_start = audio.size(0) - self.segment_length
+            audio_start = random.randint(0, max_audio_start)
+            audio = audio[audio_start:audio_start+self.segment_length]
+        else:
+            audio = torch.nn.functional.pad(audio, (0, self.segment_length - audio.size(0)), 'constant').data
+
+        mel = self.get_mel(audio)
+        audio = utils.mu_law_encode(audio / utils.MAX_WAV_VALUE, self.mu_quantization)
```

Comment:
This does not return a one-hot representation of the audio. This simply returns the audio quantized.

Reply:
Quantized audio is the sparse version of the one-hot representation of audio.  I guess it could be confusing if someone wanted to use a dense one-hot representation of audio, but I'm more worried about differentiating this code from code where the sample values are used as actual values, as is often used with Mixture of Logistics sampling.  So I'm going to leave it for now and we can change it if more audio representations come along and it causes confusion (I think we should discourage dense one-hot representations anyway).